sympy/utilities/lambdify.py

```diff
@@ -907,8 +908,17 @@ def _lambdifygenerated(x):
     # mtime has to be None or else linecache.checkcache will remove it
     linecache.cache[filename] = (len(funcstr), None, funcstr.splitlines(True), filename) # type: ignore
 
+    # Remove the entry from the linecache when the object is garbage collected
+    def cleanup_linecache(filename):
+        def _cleanup():
+            if filename in linecache.cache:
+                del linecache.cache[filename]
+        return _cleanup
+
     func = funclocals[funcname]
 
+    weakref.finalize(func, cleanup_linecache(filename))
```

Comment:
I think you want the weakref to be on the `func.__code__`

Reply:
Can you explain more why that would do anything? I would have thought that would be equivalent.

Reply:
The function can be deleted, but the code object can outlive it, eg tracebacks references the code object not the function